openwisp_firmware_upgrader/upgraders/openwrt.py

```diff
@@ -195,14 +197,34 @@ def _verify_device_uuid(self):
 
     def upgrade(self, image):
         self._test_connection()
+        self._check_cancellation()
         self._verify_device_uuid()
+        self._check_cancellation()
         checksum = self._test_checksum(image)
+        self._check_cancellation()
         remote_path = self.get_remote_path(image)
-        self.upload(image.file, remote_path)
+        self.upload(image, remote_path)
+        self._check_cancellation()
         self._test_image(remote_path)
         self._reflash(remote_path)
         self._write_checksum(checksum)
 
+    def _check_cancellation(self):
+        """
+        Check if the upgrade operation has been cancelled.
+        """
+        self.upgrade_operation.refresh_from_db()
+
```

Comment:
Let's not add so many blank lines please. Follow the current code structure.

If you have problem reading the code you'd better off increasing line-height.